### DISCUSSION TITLE
automatically replace null values with a default?

### DISCUSSION THREAD

**Discussion by tksharpless** [nlohmann/json#2860] [Q&A] [answered]
Some Json packages allow to specify a default value to be returned in case a named field is missing or null.  Is there an easy way to have that in nlohmann?  I really appreciate the good automation of return type, but would be willing to call a function to avoid having to check every time for null.  However 99% of the time the default could be a static constant.

**Comment by nlohmann** [comment, marked as answer]
The function is called `value`, see https://json.nlohmann.me/api/basic_json/value/

**Reply by tksharpless** [reply]
Thanks for the quick answer.  And for the great package!

**Reply by tksharpless** [reply]
Unfortunately it is not working for me.  Either
  FLnom_mm = par["theoreticalfocallength"];  
or
  FLnom_mm = par.value("theoreticalfocallength", 0.);
throws a value type error when that field is absent.   What am I doing wrong?

**Reply by nlohmann** [reply]
Can you share the JSON and your code?

**Reply by tksharpless** [reply]
The whole json file is very large and partly proprietary.  But here is the section call 'par' above:

> 					"params": {
						"projection": "rectilinear",
						"focallength": 1.6e1,
						"theoreticalfocallength": null,
						"croprectanglesize": null,
						"cropcenteroffset": {
							"longside": 0,
							"shortside": 0
						},
						"a": 0,
						"b": 0,
						"c": 0,
						"sensordiagonal": 2.82065595207919984e1,
						"mirrored": false
					},
As you can see, the offending field is present with an explicit value of null. 
The subsequent field, 'croprectanglesize' has the same value, but this code handles it correctly

>     myJsonArray crsz = par["croprectanglesize"];
    if( crsz.empty() ){
        cropflong = 1.0;
        cropfshort = 1.0;
    } else {
        cropflong = crsz[0];
        cropfshort = crsz[1];
    }

myJsonArray is a typedef for json

**Reply by tksharpless** [reply]
The following code works

> 	FLnom_mm 
		= par["theoreticalfocallength"].empty() 
		? 0
		: par["theoreticalfocallength"];

so the problem is not that the null value was not parsed correctly; rather it seems that value() does not handle the case of a null  scalar.

**Reply by tksharpless** [reply]
BTW I'm using version 3.9.1 from the vcpkg repository.

**Reply by tksharpless** [reply]
I don't feel quite up to modifying the code for value(); but if you would like to suggest a fix I will try it

>     // using std::is_convertible in a std::enable_if will fail when using explicit conversions
    template < class ValueType, typename std::enable_if <
                   detail::is_getable<basic_json_t, ValueType>::value
                   && !std::is_same<value_t, ValueType>::value, int >::type = 0 >
    ValueType value(const typename object_t::key_type& key, const ValueType& default_value) const
    {
        // at only works for objects
        if (JSON_HEDLEY_LIKELY(is_object()))
        {
            // if key is found, return value and given default value otherwise
            const auto it = find(key);
            if (it != end())
            {
                return it->template get<ValueType>();
            }

            return default_value;
        }

        JSON_THROW(type_error::create(306, "cannot use value() with " + std::string(type_name())));
    }

**Reply by tksharpless** [reply]
What I can see with the MSVS debugger suggests this might work:

        if (it != end())
        {
            if(it.m_it.object_iterator._Ptr->_Myval.second != nullptr)
                return it->template get<ValueType>();
        }
        return default_value;

However that  might be platform dependent?

**Reply by nlohmann** [reply]
And what is the exact error message? I am struggling to understand what you are doing.

**Reply by tksharpless** [reply]
The exception is is thrown by nlohmann::detail.get_arithmetic_value().  The error message states that a null object was found when a number was expected.
I am doing nothing unusual, simply reading values from json into variables.

**Reply by nlohmann** [reply]
`value` returns the default value if the given key is not found. If it is found, but `null` is stored, you have such a type mismatch.

**Reply by tksharpless** [reply]
Correct.  
I believe that in this case value() should also return the default . 
So you may consider this a bug report on value().

**Reply by nlohmann** [reply]
That would be a breaking change, so I cannot just fix it.

**Reply by nlohmann** [reply]
So this would rather be a new function. See also https://github.com/nlohmann/json/discussions/2715

**Reply by tksharpless** [reply]
I don't understand why value() cannot simply check if the found value is null.  What specification would that violate?
Do you seriously think any user is depending on the thrown exception to detect null objects?

**Reply by tksharpless** [reply]
Anyhow I intend to 'just fix' my copy -- I promise not to share it.  
Can you suggest a better way than going straight to the iterator's _Ptr field?

**Reply by nlohmann** [reply]
> I don't understand why value() cannot simply check if the found value is null. What specification would that violate?
> Do you seriously think any user is depending on the thrown exception to detect null objects?

Yes, this is the idea behind semantic versioning.

**Reply by nlohmann** [reply]
There is a `is_null()` function that you can call on a `basic_json` value to check if the type is `null`.

**Reply by tksharpless** [reply]
OK.  So my private version of value() has
            const auto it = find(key);
            if (it != end()  && !it.m_it.object_iterator->second.is_null())
            {
			return it->template get<ValueType>();
            }
            return default_value;

That does what I want and I'm happy.
